contracts/Joint.sol

```diff
@@ -159,85 +185,302 @@ contract Joint {
+                );
+            if (swapTo == tokenA) {
+                _aBalance = _aBalance.add(outAmounts[outAmounts.length - 1]);
+            } else if (swapTo == tokenB) {
+                _bBalance = _bBalance.add(outAmounts[outAmounts.length - 1]);
+            }
+        }
+
+        (address sellToken, uint256 sellAmount) =
+            calculateSellToBalance(_aBalance, _bBalance, investedA, investedB);
+
+        (uint256 reserveA, uint256 reserveB) = getReserves();
+
+        if (sellToken == tokenA) {
+            uint256 buyAmount =
+                IUniswapV2Router02(router).getAmountOut(
+                    sellAmount,
+                    reserveA,
+                    reserveB
+                );
+            _aBalance = _aBalance.sub(sellAmount);
+            _bBalance = _bBalance.add(buyAmount);
+        } else if (sellToken == tokenB) {
+            uint256 buyAmount =
+                IUniswapV2Router02(router).getAmountOut(
+                    sellAmount,
+                    reserveB,
+                    reserveA
+                );
+            _bBalance = _bBalance.sub(sellAmount);
+            _aBalance = _aBalance.add(buyAmount);
+        }
+
+        _aBalance = _aBalance.add(balanceOfA());
+        _bBalance = _bBalance.add(balanceOfB());
+    }
+
+    function estimatedTotalAssetsInToken(address token)
+        external
+        view
+        returns (uint256 _balance)
+    {
+        require(token == tokenA || token == tokenB);
+        if (token == tokenA) {
+            (_balance, ) = estimatedTotalAssetsAfterBalance();
         } else {
-            distributeProfit();
+            (, _balance) = estimatedTotalAssetsAfterBalance();
         }
     }
 
-    function createLP() internal {
-        IUniswapV2Router02(router).addLiquidity(
-            tokenA,
-            tokenB,
-            balanceOfA(),
-            balanceOfB(),
-            0,
-            0,
-            address(this),
-            now
-        );
+    event Ratios(uint256 tokenA, uint256 tokenB, string description);
+    event SellToBalance(
+        address sellToken,
+        uint256 sellAmount,
+        uint256 buyAmount
+    );
+
+    function calculateSellToBalance(
+        uint256 currentA,
+        uint256 currentB,
+        uint256 startingA,
+        uint256 startingB
+    ) internal view returns (address _sellToken, uint256 _sellAmount) {
+        if (startingA == 0 || startingB == 0) return (address(0), 0);
+
+        (uint256 ratioA, uint256 ratioB) =
+            getRatios(currentA, currentB, startingA, startingB);
+
+        if (ratioA == ratioB) return (address(0), 0);
+
+        uint256 numerator;
+        uint256 denominator;
+        uint256 precision;
+        uint256 exchangeRate;
+        if (ratioA > ratioB) {
+            _sellToken = tokenA;
+            precision = 10**uint256(IERC20Extended(tokenA).decimals());
+            (, exchangeRate) = getSpotExchangeRates();
```

Comment:
thought: 

this does not take into account slippage. is it relevant?

as we will be moving relevant $$$, it may be worth it to think whether or not there is any way to take it into account

Reply:
I've tried to solve the rebalance formula taking into account slippage, but i haven't cracked it. Might be another set of eyes.